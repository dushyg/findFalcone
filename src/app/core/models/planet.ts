export interface IPlanet {
	name : string;
	distance : number;	
	includedInSearch : boolean;
}
