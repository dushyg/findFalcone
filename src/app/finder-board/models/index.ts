export { IFalconAppState } from './falconApp.state';
export { RawPlanet, IPlanet } from './planet';
export { RawVehicle, IVehicle } from './vehicle';
export { default as PlanetChange } from './planetChange';
export { default as VehicleChange } from './vehicleChange';
export { ISearchAttempt } from './searchAttempt';
