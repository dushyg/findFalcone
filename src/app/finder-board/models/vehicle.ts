// tslint:disable: indent
export interface IVehicle {
  name: string;
  totalNumUnits: number;
  maxDistance: number;
  speed: number;
  availNumUnits: number;
}

export interface RawVehicle {
  id?: number;
  name: string;
  total_no: number;
  max_distance: number;
  speed: number;
}
