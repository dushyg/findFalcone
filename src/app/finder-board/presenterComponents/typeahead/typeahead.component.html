<div class="typeahead-container" (focusout)="hideResultsList()">
  <div class="typeadhead-wrapper">
    <input
      type="text"
      class="typeAhead"
      placeholder="Type to search..."
      [formControl]="inputTextControl"
      (focus)="showResultsList()"
    />
    <div class="resultList" *ngIf="doShowResults">
      <ul>
        <li class="resultItem" (click)="planetSelectHandler('Select')">
          None
        </li>
        <li
          class="resultItem"
          (click)="planetSelectHandler(planet.name)"
          *ngFor="let planet of filteredSourceArray"
        >
          {{ planet.name }}
        </li>
      </ul>
    </div>
  </div>
</div>
