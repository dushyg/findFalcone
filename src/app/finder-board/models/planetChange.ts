import { IPlanet } from './planet';

export default class PlanetChange {
  constructor(public widgetId: number, public newPlanetName: string) {}
}
