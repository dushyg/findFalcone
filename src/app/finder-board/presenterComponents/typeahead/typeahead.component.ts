import {
  Component,
  Input,
  Output,
  EventEmitter,
  OnInit,
  OnDestroy,
  OnChanges,
  SimpleChanges,
} from '@angular/core';
import { IPlanet } from 'src/app/finder-board/models/planet';
import { FormControl } from '@angular/forms';
import { Observable } from 'rxjs';
import { takeWhile } from 'rxjs/operators';

@Component({
  selector: 'app-typeahead',
  templateUrl: './typeahead.component.html',
  styleUrls: ['./typeahead.component.scss'],
})
export class TypeaheadComponent implements OnInit, OnChanges, OnDestroy {
  @Input() sourceArray: IPlanet[];
  @Input() resetTypeAhead$: Observable<void>;
  @Output() itemSelected = new EventEmitter<IPlanet>();

  public filteredSourceArray: IPlanet[];
  public inputTextControl = new FormControl('');
  public doShowResults = false;
  private lastSelection = '';
  private isComponentActive = true;

  ngOnInit(): void {
    this.filteredSourceArray = this.sourceArray;

    this.inputTextControl.valueChanges.subscribe(this.performSearch);

    this.resetTypeAhead$
      .pipe(takeWhile(() => this.isComponentActive))
      .subscribe(() => {
        // Need to wrap call to reset in setTimeout because this subscription happens before
        // the sourceArray is updated via inputs. Without setTimeout the reset function will
        // set the filteredSourceArray to old sourceArray.
        setTimeout(() => {
          this.reset();
        }, 0);
      });
  }

  ngOnChanges(changes: SimpleChanges): void {
    console.log('TypeAhead component onchanges', changes);
  }

  private reset() {
    console.log('Typeahead component reset, sourcearray', this.sourceArray);
    this.filteredSourceArray = this.sourceArray;
    this.lastSelection = '';
    this.inputTextControl.setValue('', { emitEvent: false });
  }

  private performSearch = (value) => {
    const trimmedValue = value.trim().toLowerCase();
    if (!trimmedValue) {
      this.filteredSourceArray = this.sourceArray;
    }

    this.filteredSourceArray = this.sourceArray.filter((item) => {
      return item.name.toLowerCase().includes(trimmedValue);
    });
  };

  public planetSelectHandler(planetName) {
    let selectedPlanet = this.filteredSourceArray.find((planet) => {
      return planet.name === planetName;
    });

    if (planetName === 'Select') {
      selectedPlanet = { name: 'Select', distance: 0 } as IPlanet;
      this.lastSelection = '';
      this.inputTextControl.setValue('', { emitEvent: false });
    } else {
      this.lastSelection = selectedPlanet.name;
      this.inputTextControl.setValue(this.lastSelection);
    }

    this.itemSelected.emit(selectedPlanet);
  }

  public showResultsList() {
    this.doShowResults = true;
  }

  public hideResultsList() {
    // focusout event on the wrapper div is fired first,
    // hence we need to call this after some delay so that lastSelection field
    // has latest value
    setTimeout(this.hideResults, 250);
  }

  private hideResults = () => {
    this.doShowResults = false;

    const currentText = this.inputTextControl.value;
    if (currentText === this.lastSelection) {
      return;
    }

    this.inputTextControl.setValue(this.lastSelection, { emitEvent: false });
  };

  ngOnDestroy(): void {
    this.isComponentActive = false;
  }
}
