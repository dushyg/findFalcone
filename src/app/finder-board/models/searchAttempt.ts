import { IVehicle } from "./vehicle";
import { IPlanet } from "./planet";

export interface ISearchAttempt {
  searchedPlanet: string;
  vehicleUsed: string;
}
