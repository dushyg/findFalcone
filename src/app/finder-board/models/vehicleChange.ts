import { IVehicle } from './vehicle';

export default class VehicleChange {
  constructor(public widgetId: number, public newVehicleName) {}
}
