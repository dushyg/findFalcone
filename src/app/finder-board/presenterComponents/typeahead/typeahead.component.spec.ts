import { ComponentFixture, TestBed } from '@angular/core/testing';
import { TypeaheadComponent } from './typeahead.component';
import { ReactiveFormsModule } from '@angular/forms';
import { of } from 'rxjs';
import { IPlanet } from '../../models/planet';
import { By } from '@angular/platform-browser';
import { ChangeDetectorRef, Type } from '@angular/core';

describe('TypeAhead Component', () => {
  let fixture: ComponentFixture<TypeaheadComponent>;
  let component: TypeaheadComponent;
  let planets: IPlanet[];

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [TypeaheadComponent],
      imports: [ReactiveFormsModule],
    });

    fixture = TestBed.createComponent(TypeaheadComponent);
    component = fixture.componentInstance;

    planets = [
      { name: 'Donlon', distance: 100 },
      { name: 'Enchai', distance: 200 },
      { name: 'Jebing', distance: 300 },
      { name: 'Sapir', distance: 400 },
    ];
  });

  it('is created', () => {
    component.sourceArray = [];
    component.resetTypeAhead$ = of(null);

    fixture.detectChanges();

    expect(fixture.debugElement.componentInstance).toBeTruthy();
  });

  it('renders typeahaed input text box', () => {
    component.sourceArray = [];
    component.resetTypeAhead$ = of(null);

    fixture.detectChanges();

    expect(
      fixture.debugElement.query(By.css('input.typeAhead[type=text]'))
    ).toBeTruthy();
  });

  it('shows result list if input text box recieves focus', () => {
    component.sourceArray = planets;
    component.resetTypeAhead$ = of(null);

    fixture.detectChanges();

    const inputDebugElement = fixture.debugElement.query(
      By.css('input.typeAhead[type=text]')
    );
    inputDebugElement.triggerEventHandler('focus', null);

    fixture.detectChanges();

    const resultListUlDebugElement = fixture.debugElement.query(
      By.css('div.resultList ul')
    );

    expect(resultListUlDebugElement).toBeTruthy();
  });

  it('result list has expected values', () => {
    component.sourceArray = planets;
    component.resetTypeAhead$ = of(null);

    fixture.detectChanges();

    const inputDebugElement = fixture.debugElement.query(
      By.css('input.typeAhead[type=text]')
    );
    inputDebugElement.triggerEventHandler('focus', null);

    fixture.detectChanges();

    const resultListLiElements = (fixture.nativeElement as HTMLElement).querySelectorAll(
      'div.resultList ul li'
    );
    expect(resultListLiElements).toBeTruthy();
    expect(resultListLiElements.length).toBe(planets.length + 1);

    resultListLiElements.forEach((li, index) => {
      if (index > 0) {
        expect((li as HTMLElement).textContent).toContain(
          planets[index - 1].name
        );
      } else {
        expect((resultListLiElements[0] as HTMLElement).textContent).toContain(
          'None'
        );
      }
    });
  });

  it('filters based on input text', () => {
    component.sourceArray = planets;
    component.resetTypeAhead$ = of(null);

    fixture.detectChanges();

    const inputElement: HTMLInputElement = fixture.nativeElement.querySelector(
      'input.typeAhead[type=text]'
    );
    inputElement.dispatchEvent(new Event('focus'));
    inputElement.value = 'Don';
    inputElement.dispatchEvent(new Event('input'));

    // could have also triggered this via debugelement triggerEvenHandler like this
    /* const inputDebugElement = fixture.debugElement.query(
      By.css('input.typeAhead[type=text]')
      );
      inputDebugElement.triggerEventHandler('input', {target : inputDebugElement.nativeElement}); */
    fixture.detectChanges();

    const resultListLiElements = (fixture.nativeElement as HTMLElement).querySelectorAll(
      'div.resultList ul li'
    );

    expect(resultListLiElements).toBeTruthy();
    expect(resultListLiElements.length).toEqual(2);

    expect(resultListLiElements[0].textContent).toContain('None');
    expect(resultListLiElements[1].textContent).toContain('Donlon');
  });

  it('filters even for partial planet name in input text', () => {
    component.sourceArray = planets;
    component.resetTypeAhead$ = of(null);

    fixture.detectChanges();

    const inputElement: HTMLInputElement = fixture.nativeElement.querySelector(
      'input.typeAhead[type=text]'
    );

    const inputDebugElement = fixture.debugElement.query(
      By.css('input.typeAhead[type=text]')
    );
    inputDebugElement.triggerEventHandler('focus', null);
    inputDebugElement.nativeElement.value = 'i';
    inputDebugElement.triggerEventHandler('input', {
      target: inputDebugElement.nativeElement,
    });
    fixture.detectChanges();

    const resultListLiElements = (fixture.nativeElement as HTMLElement).querySelectorAll(
      'div.resultList ul li'
    );

    expect(resultListLiElements).toBeTruthy();
    expect(resultListLiElements.length).toEqual(4);
  });

  it('emits selected planet when result item is clicked', () => {
    component.sourceArray = planets;
    component.resetTypeAhead$ = of(null);

    fixture.detectChanges();

    const inputElement: HTMLInputElement = fixture.nativeElement.querySelector(
      'input.typeAhead[type=text]'
    );

    const inputDebugElement = fixture.debugElement.query(
      By.css('input.typeAhead[type=text]')
    );
    inputDebugElement.triggerEventHandler('focus', null);
    fixture.detectChanges();

    const resultListLiElements = (fixture.nativeElement as HTMLElement).querySelectorAll(
      'div.resultList ul li'
    );

    const spy = jest.spyOn(component.itemSelected, 'emit');
    resultListLiElements[1].dispatchEvent(new Event('click'));
    fixture.detectChanges();
    expect(spy).toHaveBeenCalledWith(planets[0]);
    // this gives the value emitted by the itemSelected event emitter
    // console.log(spy.mock.calls[0][0]);
  });
});
