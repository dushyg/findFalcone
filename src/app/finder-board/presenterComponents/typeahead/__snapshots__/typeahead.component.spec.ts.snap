// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeAhead Component shows result list if input text box recieves focus 1`] = `
<app-typeahead
  __ngContext__={[Function LRootView]}
  changeDetector={[Function ViewRef]}
  doShowResults={[Function Boolean]}
  filteredSourceArray={[Function Array]}
  hideResults={[Function Function]}
  inputTextControl={[Function FormControl]}
  isComponentActive={[Function Boolean]}
  itemSelected={[Function EventEmitter]}
  lastSelection=""
  performSearch={[Function Function]}
  resetTypeAhead$={[Function Observable]}
  sourceArray={[Function Array]}
>
  <div
    class="typeahead-container"
  >
    <div
      class="typeadhead-wrapper"
    >
      <label
        for="typeahead_input_widget_"
      >
        Destination 
      </label>
      <input
        class="typeAhead ng-untouched ng-pristine ng-valid"
        id="typeahead_input_widget_"
        placeholder="Type to search..."
        type="text"
      />
      <div
        class="resultList"
      >
        <ul>
          <li
            class="resultItem"
          >
             None 
          </li>
          <li
            class="resultItem"
          >
             Donlon 
          </li>
          <li
            class="resultItem"
          >
             Enchai 
          </li>
          <li
            class="resultItem"
          >
             Jebing 
          </li>
          <li
            class="resultItem"
          >
             Sapir 
          </li>
          
        </ul>
      </div>
      
    </div>
  </div>
</app-typeahead>
`;
